presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroupManager.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.resourceGroups;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.QueryExecution;
+import com.facebook.presto.execution.QueryQueueDefinition;
+import com.facebook.presto.execution.QueryQueueManager;
+import com.facebook.presto.execution.SqlQueryManagerStats;
+import com.facebook.presto.execution.resourceGroups.ResourceGroup.RootResourceGroup;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.sql.tree.Statement;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.spi.StandardErrorCode.USER_ERROR;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.units.DataSize.Unit.BYTE;
+
+@ThreadSafe
+public class ResourceGroupManager
+        implements QueryQueueManager
+{
+    private static final Logger log = Logger.get(ResourceGroupManager.class);
+    private final ScheduledExecutorService refreshExecutor = Executors.newSingleThreadScheduledExecutor();
+    private final List<RootResourceGroup> rootGroups = new CopyOnWriteArrayList<>();
+    private final ConcurrentMap<ResourceGroupId, ResourceGroup> groups = new ConcurrentHashMap<>();
+    private final List<ResourceGroupSelector> selectors;
+
+    @Inject
+    public ResourceGroupManager(List<? extends ResourceGroupSelector> selectors)
+    {
+        this.selectors = ImmutableList.copyOf(selectors);
+    }
+
+    @Override
+    public boolean submit(Statement statement, QueryExecution queryExecution, Executor executor, SqlQueryManagerStats stats)
+    {
+        List<QueryQueueDefinition> queues = selectQueues(statement, queryExecution.getSession());
+        checkState(!queues.isEmpty(), "Selector returned empty list of queues");
```

Comment:
I'd move this to `selectQueues` so you can include the name of the bad selector in the error message.